clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -1354,6 +1402,274 @@ void csrgemm(descriptor_ptr desc, oneapi::mkl::transpose trans_a,
+/// \param [out] val_c An array containing the non-zero elements of the matrix
+/// C.
+/// \param [in] row_ptr_c An array of length row number + 1.
+/// \param [out] col_ind_c An array containing the column indices in index-based
+/// numbering.
+template <typename T>
+void csrgemm2(descriptor_ptr desc, int m, int n, int k, const T *alpha,
+              const std::shared_ptr<matrix_info> info_a, const T *val_a,
+              const int *row_ptr_a, const int *col_ind_a,
+              const std::shared_ptr<matrix_info> info_b, const T *val_b,
+              const int *row_ptr_b, const int *col_ind_b, const T *beta,
+              const std::shared_ptr<matrix_info> info_d, const T *val_d,
+              const int *row_ptr_d, const int *col_ind_d,
+              const std::shared_ptr<matrix_info> info_c, T *val_c,
+              const int *row_ptr_c, int *col_ind_c) {
+  using Ty = typename ::dpct::detail::lib_data_traits_t<T>;
+  sycl::queue &queue = desc->get_queue();
+  auto alpha_value =
+      dpct::detail::get_value(reinterpret_cast<const Ty *>(alpha), queue);
+  auto beta_value =
+      dpct::detail::get_value(reinterpret_cast<const Ty *>(beta), queue);
+
+  detail::csrgemm2_args_info args(
+      m, n, k, info_a, val_a, row_ptr_a, col_ind_a, info_b, val_b, row_ptr_b,
+      col_ind_b, info_d, val_d, row_ptr_d, col_ind_d, info_c, row_ptr_c);
+  auto &info = desc->get_csrgemm2_info_map()[args];
+  if (info.is_empty()) {
+    throw std::runtime_error("csrgemm2_nnz is not invoked previously.");
+  }
+
+  info.matrix_handle_c.set_matrix_data<Ty>(m, n, info_c->get_index_base(),
+                                           row_ptr_c, col_ind_c, val_c);
+  sycl::event e;
+#ifndef DPCT_USM_LEVEL_NONE
+  e =
+#endif
+      oneapi::mkl::sparse::omatadd(
+          queue, oneapi::mkl::transpose::nontrans,
+          oneapi::mkl::transpose::nontrans, alpha_value,
+          info.matrix_handle_c1.get_handle(), beta_value,
+          info.matrix_handle_d.get_handle(), info.matrix_handle_c.get_handle(),
+          oneapi::mkl::sparse::omatadd_alg::default_alg,
+          info.omatadd_desc.get_handle());
+
+  info.matrix_handle_c1.add_dependency(e);
+  info.matrix_handle_d.add_dependency(e);
+  info.matrix_handle_c.add_dependency(e);
+  info.omatadd_desc.add_dependency(e);
+
+  std::vector<sycl::event> events;
+  events.push_back(info.matrix_handle_c1.release());
+  events.push_back(info.matrix_handle_d.release());
+  events.push_back(info.matrix_handle_c.release());
+  events.push_back(info.omatadd_desc.release());
+  events.push_back(e);
+  desc->get_csrgemm2_info_map().erase(args);
+  queue.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(events);
+    cgh.host_task([_p1 = info.row_ptr_c1, _p2 = info.col_ind_c1,
+                   _p3 = info.val_c1, _p4 = info.ws, _q = queue] {
+      ::dpct::cs::free(_p1, _q);
+      ::dpct::cs::free(_p2, _q);
+      ::dpct::cs::free(_p3, _q);
+      ::dpct::cs::free(_p4, _q);
+    });
+  });
+}
+
```

Comment:
Overall, this is fantastic! I think adding an example for performing `C = alpha * op(A) * op(B) + beta * op(D)` as part of oneMKL library will also benefit us!